subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy

```diff
@@ -0,0 +1,364 @@
+            def service = serviceRef.get()
+            def requiredServices = service._requiredServices
+            CompletableFuture.runAsync({
+                println "backgroundJobExecutor.executed"
+            }, requiredServices.backgroundJobExecutor).exceptionally { e ->
+                println "backgroundJobExecutor.status = \${e.cause.message}"
+            }
+
+            task check {}
+        """
+
+        when:
+        succeeds("check")
+
+        then:
+        plugin.assertBackgroundJobCompletedBeforeShutdown(output, "backgroundJobExecutor.executed")
+    }
+
+    def "background job executor runs jobs submitted at execution time"() {
+        given:
+        buildFile << """
+            import ${CompletableFuture.name}
+
+            def serviceRef = gradle.extensions.serviceRef
+            task check {
+                doLast {
+                    def service = serviceRef.get()
+                    def requiredServices = service._requiredServices
+                    CompletableFuture.runAsync({
+                        println "backgroundJobExecutor.executed"
+                    }, requiredServices.backgroundJobExecutor).exceptionally { e ->
+                        println "backgroundJobExecutor.status = \${e.cause.message}"
+                    }
+                }
+            }
+        """
+
+        when:
+        succeeds("check")
+
+        then:
+        plugin.assertBackgroundJobCompletedBeforeShutdown(output, "backgroundJobExecutor.executed")
+
+        when:
+        succeeds("check")
+
+        then:
+        plugin.assertBackgroundJobCompletedBeforeShutdown(output, "backgroundJobExecutor.executed")
+    }
+
+    def "background job failure at configuration time fails build"() {
+        given:
+        buildFile << """
+            def serviceRef = gradle.extensions.serviceRef
+            def service = serviceRef.get()
+            def requiredServices = service._requiredServices
+            requiredServices.backgroundJobExecutor.execute {
+                 throw new RuntimeException("Background job failed!")
+            }
+
+            task check {}
+        """
+
+        when:
+        fails("check")
+
+        then:
+        errorOutput.contains("java.lang.RuntimeException: Background job failed!")
+    }
+
+    def "background job failure at execution time fails build"() {
+        given:
+        buildFile << """
+            def serviceRef = gradle.extensions.serviceRef
+            def service = serviceRef.get()
+            def requiredServices = service._requiredServices
+
+            task check {
+                doLast {
+                    requiredServices.backgroundJobExecutor.execute {
+                         throw new RuntimeException("Background job failed!")
+                    }
+                }
+            }
+        """
+
+        when:
+        fails("check")
+
+        then:
+        errorOutput.contains("java.lang.RuntimeException: Background job failed!")
```

Comment:
I'd like to see more of what the user experience is here for this kind of failure.